umodbus/serial.py

```diff
@@ -231,63 +299,148 @@ def _uart_read_frame(self, timeout: Optional[int] = None) -> bytearray:
-                      count: bool) -> bytes:
+        :param      slave_addr:      The slave address
+        :type       slave_addr:      int
+        :param      function_code:   The function code
+        :type       function_code:   int
+        :param      exception_code:  The exception code
+        :type       exception_code:  int
+
+        :returns:   Request response - None for a synchronous server, or
+                    an awaitable for an asynchronous server due to AsyncRequest
+        :rtype      Awaitable, optional
         """
-        Send a modbus message and receive the reponse.
+        modbus_pdu = functions.exception_response(
+            function_code=function_code,
+            exception_code=exception_code)
+        return self._send(modbus_pdu=modbus_pdu, slave_addr=slave_addr)
 
-        :param      modbus_pdu:  The modbus Protocol Data Unit
-        :type       modbus_pdu:  bytes
-        :param      slave_addr:  The slave address
-        :type       slave_addr:  int
-        :param      count:       The count
-        :type       count:       bool
 
-        :returns:   Validated response content
-        :rtype:     bytes
+class Serial(CommonRTUFunctions, CommonModbusFunctions):
+    """Modbus Serial/RTU client"""
+
+    def _exit_read(self, response: bytearray) -> bool:
         """
-        # flush the Rx FIFO
-        self._uart.read()
+        Return on modbus read error
 
-        self._send(modbus_pdu=modbus_pdu, slave_addr=slave_addr)
+        :param      response:    The response
+        :type       response:    bytearray
 
-        return self._validate_resp_hdr(response=self._uart_read(),
-                                       slave_addr=slave_addr,
-                                       function_code=modbus_pdu[0],
-                                       count=count)
+        :returns:   State of basic read response evaluation,
+                    True if entire response has been read
+        :rtype:     bool
+        """
+        response_len = len(response)
+        if response_len >= 2 and response[1] >= Const.ERROR_BIAS:
+            if response_len < Const.ERROR_RESP_LEN:
+                return False
+        elif response_len >= 3 and (Const.READ_COILS <= response[1] <= Const.READ_INPUT_REGISTER):
+            expected_len = Const.RESPONSE_HDR_LENGTH + 1 + response[2] + Const.CRC_LENGTH
+            if response_len < expected_len:
+                return False
+        elif response_len < Const.FIXED_RESP_LEN:
+            return False
+
+        return True
+
+    def _uart_read(self) -> bytearray:
+        """
+        Read incoming slave response from UART
+
+        :returns:   Read content
+        :rtype:     bytearray
+        """
+        response = bytearray()
+        # number of repetitions = <wait_time_in_ms> // <sleep_per_repetition>
+        repetitions = self._uart_read_timeout // self._inter_frame_delay
+
+        for _ in range(1, repetitions):
```

Comment:
see previous comment, stay with `40` for now, I'll place another PR for this requirement after merging this PR

Reply:
Should it be 40 or 120? I think I can change it back to 120 to make it easier to merge, and then after this PR is merged in it can be changed to 40 or a custom timeout later.